### ISSUE TITLE
custom login mapping doesn't work

### ISSUE THREAD

**Issue by gravuz** [RainLoop/rainloop-webmail#1235]
On Rainloop 1.10.4.183 the plugin custom-login-mapping doesn't work. After setting up the plugin i can't login in webmail by the user i set. The error is: Authentication failed.
Thanks you

**Comment by ervee** [comment]
I'm using that version and have no issue with the user mapping.

**Comment by gravuz** [comment]
yes, but if it's enable i can't login with my e-mail address or the username i set. I'm not using the whitelist

**Comment by ervee** [comment]
Please enable RainLoop logging and post your logs.

**Comment by tomas4** [comment]
The plugin does not work for me either. First i have found issue #1012 , so i added the comment there, but this one seems more appropriate, since OP problem seems identical to mine and I have the logs requested. Here goes slightly edited copy of my reply in the other issue. I hope it can help or you guys can tell me what I am doing wrong:

**The situation:**
My e-mail address is in the form of Jan.Novak@example.com, IMAP username is in the form jano1234@example.com (actual email and user differs). The e-mail server is in fact outlook.com corporate account. I have added settings for the example.com domain.

**Without using the plugin**, I am able to get into my account and read and send e-mails, if I actually use jano1234@example.com as e-mail address when adding the account to RainLoop. But in this case jano1234@example.com is also the From: address in the sent e-mails. I can set the Reply-To: to the correct address, but many e-mail clients do not respect this and even in cases when they do, the recipient still sees my email as received from nonexistent address jano1234@example.com.

**With the plugin**
So I ins
[log-2020-10-23.txt](https://github.com/RainLoop/rainloop-webmail/files/5430170/log-2020-10-23.txt)
[log-2020-10-23.txt](https://github.com/RainLoop/rainloop-webmail/files/5430177/log-2020-10-23.txt)

talled the plugin and set the rule according the hint in the window (email:user) to:
Jan.Novak@example.com:jano1234@example.com

But when I try to add the account by the address Jan.Novak@example.com with the plugin configured, I always get authentication failed, so it seems the plugin just does not work as stated in the issue title. If I use the username jano1234@example.com to create the account (but it surely is not right, since you are asked to fill e-mail, not username), then the situation is exactly the same as without the plugin as described above, i.e. wrong from: address - so the plugin has no effect either.
I also enabled logging during attempt to add the account with the Jan.Novak@example.com email and the jano1234 string is not in the logs anywhere. It seems the e-mail is being sent to the imap server as login name (email changed,plain text password or row of stars replaced by "HIDDEN"):

`[09:19:22.594][7d49b9df] INFO[DATA]: [DATE:23.10.20][OFFSET:-00][RL:1.12.1][PHP:7.3.19-1~deb10u1][IP:::1][PID:4008][nginx/1.14.2][fpm-fcgi]
[09:19:22.594][7d49b9df] INFO[DATA]: [Suhosin:off][APC:off][MB:on][PDO:mysql][Streams:tcp,udp,unix,udg,ssl,tls,tlsv1.0,tlsv1.1,tlsv1.2]
[09:19:22.594][7d49b9df] REQUEST[NOTE]: [POST] http://localhost/rainloop/?/Ajax/&q[]=/_jyaic-lxRaQGmcaiKhZfyL63GmujeLyomb_11rcN-E-sM5YtQlSRzi5jNcjhrU-plEPvwdFYwH572FEbsOV7tTnlQQj051jdR8K72qS4EmhstEmxNTEtb9h-ZTDWJ_GI6LybjkQ9k8zh809JnDC5Bs-TSWYoBZntM1-ZMyj6rYZqKpMU7eTZe7BPJLwtJ9xWI8kTFKCv3vrei6UcYW2zxVM9cx8joX5LyKWWzgSVJgrviIXnh9IjlKqhLCtN9mVc88yyDCBbODJvBk9X2H8g_4tv6W3peq4iAmS-QvnGPRtr8GzmNV0X-Lx6SN2uW6evKyC9THjt-LIPs-fkZi8uUVoVe55KbtFKTgh72iPf_2DJDwXfojgiyE9WreoIWCkV/
[09:19:22.594][7d49b9df] AJAX[NOTE]: Action: DoAccountSetup
[09:19:22.594][7d49b9df] POST[DATA]: {"Email":"Jan.Novak@upce.cz","Password":"HIDDEN","New":"1","Action":"AccountSetup","XToken":"c4e6607fb4129d203ec2292b13470009"}
[09:19:22.595][7d49b9df] PLUGIN[NOTE]: Hook: filter.login-credentials
[09:19:22.595][7d49b9df] IMAP[NOTE]: Start connection to "ssl://imap-mail.outlook.com:993"
[09:19:22.652][7d49b9df] IMAP[NOTE]: Connected (success)
[09:19:22.701][7d49b9df] IMAP[DATA]: < * OK The Microsoft Exchange IMAP4 service is ready. [QQBNADAAUABSADAAMgBDAEEAMAAxADAAMgAuAGUAdQByAHAAcgBkADAAMgAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]\r\n
[09:19:22.701][7d49b9df] IMAP[DATA]: > TAG1 CAPABILITY\r\n
[09:19:22.718][7d49b9df] IMAP[DATA]: < * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+\r\n
[09:19:22.718][7d49b9df] IMAP[DATA]: < TAG1 OK CAPABILITY completed.\r\n
[09:19:22.718][7d49b9df] IMAP[DATA]: > TAG2 AUTHENTICATE PLAIN\r\n
[09:19:22.734][7d49b9df] IMAP[DATA]: < + \r\n
[09:19:22.735][7d49b9df] IMAP[SECURE]: > HIDDEN\r\n
[09:19:23.794][7d49b9df] IMAP[DATA]: < TAG2 NO AUTHENTICATE failed.\r\n
[09:19:23.794][7d49b9df] IMAP[WARNING]: MailSo\Imap\Exceptions\NegativeResponseException: MailSo-Imap-Exceptions-NegativeResponseException (ImapClient.php ~ 1874) in /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php:1874
Stack trace:
#0 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(1951): MailSo\Imap\ImapClient->validateResponse(Array)
#1 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(281): MailSo\Imap\ImapClient->parseResponseWithValidation()
#2 /usr/share/rainloop/app/libraries/MailSo/Mail/MailClient.php(92): MailSo\Imap\ImapClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#3 /usr/share/rainloop/app/libraries/RainLoop/Model/Account.php(451): MailSo\Mail\MailClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#4 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2078): RainLoop\Model\Account->IncConnectAndLoginHelper(Object(RainLoop\Plugins\Manager), Object(MailSo\Mail\MailClient), Object(RainLoop\Config\Application))
#5 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#6 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#7 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#8 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#9 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#10 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#11 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#12 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#13 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#14 {main}
[09:19:23.794][7d49b9df] IMAP[NOTICE]: MailSo\Imap\Exceptions\NegativeResponseException: MailSo-Imap-Exceptions-NegativeResponseException (ImapClient.php ~ 1874) in /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php:1874
Stack trace:
#0 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(1951): MailSo\Imap\ImapClient->validateResponse(Array)
#1 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(281): MailSo\Imap\ImapClient->parseResponseWithValidation()
#2 /usr/share/rainloop/app/libraries/MailSo/Mail/MailClient.php(92): MailSo\Imap\ImapClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#3 /usr/share/rainloop/app/libraries/RainLoop/Model/Account.php(451): MailSo\Mail\MailClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#4 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2078): RainLoop\Model\Account->IncConnectAndLoginHelper(Object(RainLoop\Plugins\Manager), Object(MailSo\Mail\MailClient), Object(RainLoop\Config\Application))
#5 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#6 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#7 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#8 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#9 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#10 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#11 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#12 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#13 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#14 {main}

Next MailSo\Imap\Exceptions\LoginBadCredentialsException: MailSo-Imap-Exceptions-LoginBadCredentialsException (ImapClient.php ~ 312) in /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php:312
Stack trace:
#0 /usr/share/rainloop/app/libraries/MailSo/Mail/MailClient.php(92): MailSo\Imap\ImapClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#1 /usr/share/rainloop/app/libraries/RainLoop/Model/Account.php(451): MailSo\Mail\MailClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#2 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2078): RainLoop\Model\Account->IncConnectAndLoginHelper(Object(RainLoop\Plugins\Manager), Object(MailSo\Mail\MailClient), Object(RainLoop\Config\Application))
#3 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#4 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#5 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#6 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#7 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#8 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#9 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#10 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#11 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#12 {main}
[09:19:23.795][7d49b9df] INFO[ERROR]: MailSo\Imap\Exceptions\NegativeResponseException: MailSo-Imap-Exceptions-NegativeResponseException (ImapClient.php ~ 1874) in /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php:1874
Stack trace:
#0 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(1951): MailSo\Imap\ImapClient->validateResponse(Array)
#1 /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php(281): MailSo\Imap\ImapClient->parseResponseWithValidation()
#2 /usr/share/rainloop/app/libraries/MailSo/Mail/MailClient.php(92): MailSo\Imap\ImapClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#3 /usr/share/rainloop/app/libraries/RainLoop/Model/Account.php(451): MailSo\Mail\MailClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#4 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2078): RainLoop\Model\Account->IncConnectAndLoginHelper(Object(RainLoop\Plugins\Manager), Object(MailSo\Mail\MailClient), Object(RainLoop\Config\Application))
#5 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#6 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#7 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#8 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#9 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#10 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#11 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#12 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#13 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#14 {main}

Next MailSo\Imap\Exceptions\LoginBadCredentialsException: MailSo-Imap-Exceptions-LoginBadCredentialsException (ImapClient.php ~ 312) in /usr/share/rainloop/app/libraries/MailSo/Imap/ImapClient.php:312
Stack trace:
#0 /usr/share/rainloop/app/libraries/MailSo/Mail/MailClient.php(92): MailSo\Imap\ImapClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#1 /usr/share/rainloop/app/libraries/RainLoop/Model/Account.php(451): MailSo\Mail\MailClient->Login('Jan.Novak@...', 'HIDDEN', '', true, false)
#2 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2078): RainLoop\Model\Account->IncConnectAndLoginHelper(Object(RainLoop\Plugins\Manager), Object(MailSo\Mail\MailClient), Object(RainLoop\Config\Application))
#3 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#4 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#5 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#6 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#7 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#8 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#9 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#10 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#11 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#12 {main}

Next RainLoop\Exceptions\ClientException: AuthError[102] in /usr/share/rainloop/app/libraries/RainLoop/Actions.php:2097
Stack trace:
#0 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2329): RainLoop\Actions->CheckMailConnection(Object(RainLoop\Model\Account), true)
#1 /usr/share/rainloop/app/libraries/RainLoop/Actions.php(2861): RainLoop\Actions->LoginProcess('Jan.Novak@...', 'HIDDEN', '', '', false, true)
#2 /usr/share/rainloop/app/libraries/RainLoop/ServiceActions.php(172): RainLoop\Actions->DoAccountSetup()
#3 /usr/share/rainloop/app/libraries/RainLoop/Service.php(146): RainLoop\ServiceActions->ServiceAjax('')
#4 /usr/share/rainloop/app/libraries/RainLoop/Service.php(56): RainLoop\Service->localHandle()
#5 /usr/share/rainloop/app/libraries/RainLoop/Service.php(79): RainLoop\Service->__construct()
#6 /usr/share/rainloop/app/handle.php(94): RainLoop\Service::Handle()
#7 /usr/share/rainloop/include.php(228): include('/usr/share/rain...')
#8 /usr/share/rainloop/index.php(13): include('/usr/share/rain...')
#9 {main}
[09:19:23.795][7d49b9df] AJAX[DATA]: {"Action":"AccountSetup","Result":false,"ErrorCode":102,"ErrorMessage":"","ErrorMessageAdditional":"","Time":1213}
[09:19:23.796][7d49b9df] INFO[MEMORY]: Memory peak usage: 2MB
[09:19:23.796][7d49b9df] INFO[TIME]: Time delta: 1.2145619392395
[09:19:23.797][7d49b9df] IMAP[NOTE]: Disconnected from "ssl://imap-mail.outlook.com:993" (success)
`

**Comment by FreemoX** [comment]
A little late to the party, but this is still the top result for this issue, so this is to contribute to a clear answer towards this.

The plugin allows the admin to define two values; a nice friendly email people can use to log in with, and a corresponding ugly mx server username. These are separated by a colon ( : ) respectively.
Now, on my mx server, I use arbitrary usernames. Let's say mine is 'mydomain4', and my domain is 'mydomain.com'.
In order to login normally, I'd have to login with 'mydomain4@mydomain.com'. But this plugin allows me to predefine 'hello@mydomain.com' to login with instead. 

**TL;DR**
Email I need to log in with = `mydomain4@mydomain.com`
My domain = `mydomain.com`
MX Server username = `mydomain4`
Email I want to log in with = `hello@mydomain.com`

Login mapping to use = `hello@mydomain.com:mydomain4`

I made this pull request #2147 to append a simple usage explanation to the readme/tooltip for the plugin
